lib/lib.bash

```diff
@@ -0,0 +1,363 @@
+  local -a options=()
+  local key
+
+  for key in tag service repository; do
+    if test -n "$(get-config "$key")"; then
+      options+=("$key")
+    fi
+  done
+
+  local count="${#options[@]}"
+  if test "$count" -gt 0; then
+    # First, notify the user of the issue (respecting plurals)
+    if test "$count" -eq 1; then
+      error "An image option was encountered as a top-level configuration item."
+    else
+      error "Some image options were encountered as top-level configuration items."
+    fi
+
+    # Blank line to match common image format
+    error
+
+    # Now, dump all the configuration keys that we believe are incorrect and indicate
+    # where they should live.
+    error "To remedy, please move the invalid configuration"
+    error "    image:"
+    for key in "${options[@]}"; do
+      error "      ${key}: $(get-config "$key")"
+    done
+
+    failed=$((failed + 1))
+  fi
+
+  # Check for the presence of the required from option.
+  if test -z "$(get-config from)"; then
+    error "The required from option is not present."
+    error
+    error "This plugin requires the from option to know which path to extract from the Docker"
+    error "image."
+
+    failed=$((failed + 1))
+  fi
+
+  # If we encountered any failures, then it's time to bail - the configuration isn't
+  # trustworthy.
+  if test "$failed" -gt 0; then
+    fail-build
+  fi
+}
+
+# USAGE:
+#   generate-tag
+#
+# Generates a tag in one of two forms:
+# 1. <pipeline slug>-build-<build number>
+# 2. <pipeline slug>-<service>-build-<build number>
+#
+# The former option is chosen if there is no service configuration provided, and the
+# latter is designed to mimic the Buildkite Docker Compose plugin's (see link below).
+#
+# Note that this function should not be called directly in the plugin hook. See the
+# image-name function instead.
+#
+# cf. https://github.com/buildkite-plugins/docker-compose-buildkite-plugin/blob/398a7ff32422d2911a0c32c71d8e5e2749cdf41d/lib/shared.bash#L191-L203
+generate-tag() {
+  local service
+  service="$(get-config image-service)"
+
+  # $separator will be empty if $service is, and "-" otherwise
+  local separator="${service:+-}"
+
+  echo "${BUILDKITE_PIPELINE_SLUG}-${service}${separator}build-${BUILDKITE_BUILD_NUMBER}"
+}
+
+# USAGE:
+#   image-tag
+#
+# Returns the tag to use for this image. It respects the image configuration option, but
+# knows how to generate an image tag otherwise.
+#
+# Note that this function should not be called directly in the plugin hook. See the
+# image-name function instead.
+image-tag() {
+  # The image tag option supersedes the auto-generated tag
+  local tag
+  tag="$(get-config image-tag)"
+
+  echo "${tag:-$(generate-tag)}"
+}
+
+# USAGE:
+#   generate-name
+#
+# Generates a name for the image based on configuration. It will use the configuration to determine which tag to apply, and then attemps to look up the repository
```

Comment:
Nit-pick: You seem to be very consistent on your line lengths everywhere except here.

Reply:
Goof on my part. I'll fix it.